proposals/0000-constraint-based-arrow-notation.md

```diff
@@ -0,0 +1,430 @@
+
+     ```haskell
+     keyed :: (Eq k, Applicative m) => Rule m (e, k, a) b -> Rule m (e, Map k a) (Map k b)
+     ```
+
+     Making this work with the old notation only requires adding an extra level of nesting in the first argument, but making it work with the new notation requires a dramatically more complicated type. Under this proposal, it would work in arrow notation as-written with no changes at all.
+
+In other words, the original rules were carefully designed to align with the types programmers *naturally* write, allowing ordinary arrow combinators to be used as control operators. The current desugaring used by GHC falls flat precisely because it does *not* align with those types, making it significantly less useful (so much so that users seem confused about the notation’s purpose).
+
+## Proposed Change Specification
+
+To restore the spirit of the original system, I propose the following modifications to GHC’s implementation of arrow notation.
+
+### Typechecking
+
+Both the original and current system [are specified][Paterson04] using judgments of the following form:
+
+```
+G;D |-a c :: stk --> t
+```
+
+These judgments should be read as follows: under contexts `G` and `D` containing the types of non-local and local variables, respectively, the command `c` accepts a stack of type `stk` and returns a value of type `t` in some arrow `a`. The special `-->` arrow is *not* a Haskell type, but part of the syntax of the `|-a` judgment itself. The `stk` term is a list of (mono-)types, and it is an “input” to the judgment for the purposes of an algorithmic implementation.
+
+Here is an example of an existing typechecking rule, this one for command application:
+
+```
+G;ys |-a cmd :: (t1, stk) --> t2
+G,xs |- exp :: t1
+ys \subseteq xs
+--------------------------------
+G;xs |-a cmd exp :: stk --> t2
+```
+
+I do not propose any radical changes to the structure of the `|-a` judgment or its rules, but I suggest the following key modifications:
+
+  1. Eliminate the requirement that `stk` be strictly an input to `|-a`. Instead, allow it to be an arbitrary type of kind `[Type]`. This allows `stk` to be a (potentially unsolved) metavariable rather than a concrete list, which allows GHC’s typechecker to propagate information bidirectionally using the ordinary type inference mechanism.
+
+  2. Introduce two wired-in type families:
+
+     1. `ArrowStack :: [Type] -> Type`, which converts a stack type to a tuple. Morally, it is a closed type family with the following infinitely-long definition:
+
+        ```haskell
+        type family ArrowStack stk where
+          ArrowStack '[a]       = a
+          ArrowStack '[a, b]    = (a, b)
+          ArrowStack '[a, b, c] = (a, b, c)
+          ...
+        ```
+
+        This type family is non-injective, as the RHS of the first case overlaps with all other cases.
```

Comment:
https://hackage.haskell.org/package/ghc-prim-0.5.3/docs/GHC-Tuple.html#t:Unit is a 1-ary tuple I really think we should be using more, and which would make this one injective too.

(Off topic: Rust in allowing trailing commas has `(_ ,)` as a 1-ary tuple constructor. I rather like that, and would also propose (for Haskell) making `(,)` the prefix form, so that the number of commas is the arity of the tuple rather than 1 less than the arity of the tuple. Oh, an I would make `Identity` a synonym for this `(_ ,)`.)

Reply:
Using `Unit` here would require programmers to write things like `f -< Unit x` everywhere they pass a value that isn’t a tuple instead of just `f -< x`, which clearly does not fly. That’s why it’s so crucial that `ArrowEnv` *is* injective, since it allows the input to `ArrowStack` to be inferred independently.

Reply:
OK I was wondering if we could get away with one family too, or whether the `Unit` data constructor could be added and removed automatically. It is not pressing.